### DISCUSSION TITLE
How much python do I need to know to contribute?

### DISCUSSION THREAD

**Discussion by brenobaise** [UOW-Computing/Emperor#27] [Tutorials]
### TL;DR. - Speak to nuke if you can't open links (#404)
Go through [Get Started](https://github.com/UOW-Computing/Emperor/blob/master/README.md) to set up your environment, then get an  [Issues](https://github.com/UOW-Computing/Emperor/issues).

If you don't want to code, you can also contribute by testing the bot and reporting the problems. This teaches you about the codebase.

### You need to have at least a basic idea of what these are:

Now a basic idea doesn't mean you need to know everything but you shouldn't feel like its new to you.
We have seen all of this in Software Dev I, aside from coroutines.
When I don't remember something, I personally use [w3schools](https://www.w3schools.com/python/default.asp) as a reference.

### **Python Fundamentals**
1. Data Types: variable assignment.
2. Operators logical and basic arithmetic.
3. Control Flow Statements: `If-elif-else`. 
4. Structural Pattern Matching:` Match-case`
5. Loops : `while/for loops`, understand `break`, `continue` keywords.
6. **FUNCTIONS! ideally asynchronous since Emperor relies on it. (AKA coroutines)**
7. Error Handling: `Try and Except, throws.`
8. Arrays: `Lists`,` Dictionaries`, bonus points if you know more data structures.
9. File Handling: Read and Write to files.

**With an understand of these fundamentals you are more than able to help writing code for Emperor.** 

However, I don't want to sugar coat the topic, you will struggle and the more you do the more you will learn. 
This language is the least verbose and the easiest to learn, the knowledge you learn from this carries over to other languages 100%.

### Contributing
Contributing doesn't mean you have to produce a whole module to Emperor, pick a `training bug` from [Issues](https://github.com/UOW-Computing/Emperor/issues) and try to figure out the problem. Ask for help through creating an issue and use the respective label.
Other ways of contributing to Emperor, is through documentation writing, testing the Bot, and art design.

Start Contributing : [Get Started](https://github.com/UOW-Computing/Emperor/blob/master/README.md)

### Emperor's Codebase
Emperor's codebase is becoming lengthy and eventually we will go public to get a grab of the analytics of the codebase.
What this means is the code is getting big and will 100% find concepts you don't understand.
Discord.py is the module which wraps the Discord API for python, and it is filled with code that not even I understand. So don't feel demotivated just because you don't understand what you are reading. Ask for help, but at the same time don't expect to be spoon-fed.

### Note from the editor
When you hear people say, you need to practice in order to actually learn. This is what they mean, contribute.

### Python Resources -- Under Revision
[Py4e](https://www.py4e.com/lessons)
[w3schools](https://www.w3schools.com/python/default.asp)
[freecodecamp](https://www.freecodecamp.org/news/python-programming-course/)
